### ISSUE TITLE
Only print captured console.* if a test fails.

### ISSUE THREAD

**Issue by jmoseley** [vitest-dev/vitest#6272]
### Clear and concise description of the problem

When running my tests, I see all of the console.* statements from my code. This pollutes the output and makes the test run progress hard to follow. I can use `--silent` to turn off the logs, but, in the event of a failed test, this output is useful to track down what went wrong.

### Suggested solution

I think it would be great to have a config flag that enables silent capturing of all output (like `--silent`), but outputs the captured logs for tests that fail.

### Alternative

I attempted to make a custom reporter for this, but that comes at the cost of being able to use the `default` reporter. This request feels like an extension of the `default` reporter as opposed to a new reporter altogether. Or I am lacking the context to make a custom reporter do what I would like.

### Validations

- [X] Follow our [Code of Conduct](https://github.com/vitest-dev/vitest/blob/main/CODE_OF_CONDUCT.md)
- [X] Read the [Contributing Guidelines](https://github.com/vitest-dev/vitest/blob/main/CONTRIBUTING.md).
- [X] Read the [docs](https://vitest.dev/guide/).
- [X] Check that there isn't already an issue that request the same feature to avoid creating a duplicate.

**Comment by lukeramsden** [comment]
This would be very useful - to give a concrete usecase, I use testcontainers to run integration tests against a real database, and I'm having intermittency issues where sometimes the test hooks are timing out due to taking too long and it's hard to debug as I get the logs for ALL the tests rather than just the one that timed out.

**Comment by AriPerkkio** [comment]
Recently I used this kind of approach to print logged `console.error()` calls only when tests fail. You could extend this logic to capture `.log()` calls as well:

https://github.com/AriPerkkio/vitest/blob/f11b6d8a1a688e7bd2ce271969b07ba4c982056f/test/config/test/failures.test.ts#L16-L28

**Comment by jmoseley** [comment]
@AriPerkkio that's a nice way to mimic this behavior. I'll give it a try and see how well it works.

**Comment by wilhuff** [comment]
@AriPerkkio I was able to make your approach work as well, however I ran into this error in the invocation of the logger during a test failure:
```
TypeError: Cannot read properties of undefined (reading '#stdout')
```

The following change to your code fixes this issue:
```diff
   ctx.onTestFailed(() => {
-     errors.forEach(args => original(...args))
+     errors.forEach(args => original.call(console, ...args))
   })
```

The issue is that the particular `console` object my tests were running against was failing trying to access members associated with `this` (`this.#stdout`).

This ended up being my complete solution:
```ts
import { beforeEach } from "vitest";

const loggers = ["log", "debug", "trace", "info", "warn", "error"] as const;

type LogFunction = (...args: any) => void;
type LoggedInvocation = [LogFunction, any[]];

beforeEach((ctx) => {
  const logs: LoggedInvocation[] = [];

  const original: Record<string, LogFunction> = {};
  for (const logger of loggers) {
    original[logger] = console[logger];
    console[logger] = (...args) => logs.push([original[logger], args]);
  }

  ctx.onTestFailed(() => {
    for (const [logger, data] of logs) {
      logger.call(console, ...data);
    }
  });

  return () => {
    for (const logger of loggers) {
      console[logger] = original[logger];
    }
  };
});
```

**Comment by jmoseley** [comment]
Thanks @AriPerkkio !
